File: src/geometry_triangle_properties_hard.py
"""
geometry - triangle_properties (hard)
Generated: 2026-02-22T04:05:51.510080
"""

from problem_utils import *
from svg_utils import Diagram

def generate():
    problem_type = randint(1, 5)
    
    if problem_type == 1:
        # Triangle with angle bisector theorem
        # Pick clean ratios
        ratio1 = randint(2, 5)
        ratio2 = randint(2, 5)
        while ratio2 == ratio1:
            ratio2 = randint(2, 5)
        
        multiplier = randint(2, 4)
        side1 = ratio1 * multiplier
        side2 = ratio2 * multiplier
        
        # Angle bisector divides opposite side in ratio of adjacent sides
        total_base = randint(12, 24)
        segment1 = Rational(total_base * ratio1, ratio1 + ratio2)
        segment2 = total_base - segment1
        
        d = Diagram(width=350, height=250)
        A = (0, 0)
        B = (total_base, 0)
        C = (float(segment1), 8)
        D = (float(segment1), 0)
        
        d.polygon([A, B, C], labels=['A', 'B', 'C'])
        d.line(C, D, dashed=True)
        d.segment_label(A, C, f"{side1}")
        d.segment_label(B, C, f"{side2}")
        d.segment_label(A, D, "x")
        d.segment_label(D, B, "y")
        d.angle_arc(C, A, D, label="")
        d.angle_arc(C, D, B, label="")
        
        question = f"In triangle $ABC$, $AC = {side1}$ and $BC = {side2}$. The angle bisector from $C$ meets side $AB$ at point $D$, dividing it into segments $AD = x$ and $DB = y$ where $x + y = {total_base}$. Find $x$."
        
        ans = segment1
        
        solution = steps(
            f"By the Angle Bisector Theorem, the angle bisector divides the opposite side in the ratio of the adjacent sides",
            f"Therefore, $\\frac{{AD}}{{DB}} = \\frac{{AC}}{{BC}} = \\frac{{{side1}}}{{{side2}}}$",
            f"Let $AD = x$ and $DB = y$, so $\\frac{{x}}{{y}} = \\frac{{{side1}}}{{{side2}}}$",
            f"We also know $x + y = {total_base}$",
            f"From the ratio: ${side2}x = {side1}y$, so $y = \\frac{{{side2}x}}{{{side1}}}$",
            f"Substituting: $x + \\frac{{{side2}x}}{{{side1}}} = {total_base}$",
            f"$\\frac{{{side1}x + {side2}x}}{{{side1}}} = {total_base}$",
            f"${side1 + side2}x = {total_base * side1}$",
            f"$x = {latex(ans)}$"
        )
        
        return problem(
            question=question,
            answer=ans,
            difficulty=(1650, 1750),
            topic="geometry/triangle_properties",
            solution=solution,
            image=d.render()
        )
    
    elif problem_type == 2:
        # Stewart's Theorem application
        base = randint(10, 16)
        cevian_point = randint(4, base - 4)
        other_point = base - cevian_point
        
        side_b = randint(8, 14)
        side_c = randint(8, 14)
        
        # Using Stewart's Theorem: b²·m + c²·n - d²·a = m·n·a
        # where d is cevian, a is base, m and n are segments
        # Solve for d
        m = cevian_point
        n = other_point
        a = base
        
        d_squared = (side_b**2 * m + side_c**2 * n - m * n * a) / a
        
        if d_squared > 0 and d_squared == int(d_squared):
            d_val = int(sqrt(d_squared))
            
            question = f"In triangle $ABC$, side $AB = {base}$. Point $D$ lies on $AB$ such that $AD = {m}$ and $DB = {n}$. If $AC = {side_b}$ and $BC = {side_c}$, find the length of cevian $CD$ using Stewart's Theorem."
            
            ans = d_val
            
            solution = steps(
                f"Stewart's Theorem states: $b^2 \\cdot m + c^2 \\cdot n = a(d^2 + mn)$",
                f"where $b = AC = {side_b}$, $c = BC = {side_c}$, $a = AB = {base}$, $m = AD = {m}$, $n = DB = {n}$, and $d = CD$",
                f"Substituting: ${side_b}^2 \\cdot {m} + {side_c}^2 \\cdot {n} = {base}(d^2 + {m} \\cdot {n})$",
                f"${side_b**2 * m} + {side_c**2 * n} = {base}d^2 + {base * m * n}$",
                f"${side_b**2 * m + side_c**2 * n} = {base}d^2 + {base * m * n}$",
                f"${base}d^2 = {side_b**2 * m + side_c**2 * n - base * m * n}$",
                f"$d^2 = {int(d_squared)}$",
                f"$d = {ans}$"
            )
            
            return problem(
                question=question,
                answer=ans,
                difficulty=(1700, 1800),
                topic="geometry/triangle_properties",
                solution=solution
            )
    
    elif problem_type == 3:
        # Median length formula
        a_side = randint(8, 14)
        b_side = randint(8, 14)
        c_side = randint(8, 14)
        
        # Check triangle inequality
        if a_side + b_side > c_side and a_side + c_side > b_side and b_side + c_side > a_side:
            # Median to side a: m_a = (1/2)sqrt(2b² + 2c² - a²)
            median_squared = (2 * b_side**2 + 2 * c_side**2 - a_side**2) / 4
            
            if median_squared > 0:
                median_val = sqrt(median_squared)
                
                question = f"In triangle $ABC$, $AB = {c_side}$, $AC = {b_side}$, and $BC = {a_side}$. Find the length of the median from vertex $A$ to the midpoint of side $BC$."
                
                ans = median_val
                
                solution = steps(
                    f"The median length formula from vertex $A$ to side $a$ is: $m_a = \\frac{{1}}{{2}}\\sqrt{{2b^2 + 2c^2 - a^2}}$",
                    f"where $a = BC = {a_side}$, $b = AC = {b_side}$, $c = AB = {c_side}$",
                    f"$m_a = \\frac{{1}}{{2}}\\sqrt{{2({b_side})^2 + 2({c_side})^2 - ({a_side})^2}}$",
                    f"$m_a = \\frac{{1}}{{2}}\\sqrt{{{2 * b_side**2} + {2 * c_side**2} - {a_side**2}}}$",
                    f"$m_a = \\frac{{1}}{{2}}\\sqrt{{{2 * b_side**2 + 2 * c_side**2 - a_side**2}}}$",
                    f"$m_a = {latex(ans)}$"
                )
                
                return problem(
                    question=question,
                    answer=ans,
                    difficulty=(1650, 1750),
                    topic="geometry/triangle_properties",
                    solution=solution
                )
    
    elif problem_type == 4:
        # Triangle with altitude and area relationship
        base = randint(10, 20)
        height = randint(6, 12)
        area = Rational(base * height, 2)
        
        # Find another side length
        side_b = randint(12, 18)
        # Area = (1/2) * side_b * h_b, so h_b = 2*area/side_b
        height_b = Rational(2 * area, side_b)
        
        question = f"Triangle $ABC$ has base $BC = {base}$ and the altitude from $A$ to $BC$ has length ${height}$. If side $AC = {side_b}$, find the length of the altitude from vertex $B$ to side $AC$."
        
        ans = height_b
        
        solution = steps(
            f"First, find the area of triangle $ABC$ using base $BC$ and its altitude",
            f"Area $= \\frac{{1}}{{2}} \\cdot {base} \\cdot {height} = {latex(area)}$",
            f"The area can also be calculated using side $AC = {side_b}$ as the base",
            f"If $h$ is the altitude from $B$ to $AC$, then Area $= \\frac{{1}}{{2}} \\cdot {side_b} \\cdot h$",
            f"${latex(area)} = \\frac{{{side_b}h}}{{2}}$",
            f"${2 * area} = {side_b}h$",
            f"$h = {latex(height_b)}$"
        )
        
        return problem(
            question=question,
            answer=ans,
            difficulty=(1600, 1700),
            topic="geometry/triangle_properties",
            solution=solution
        )
    
    else:
        # Heron's formula with specific sides
        a_val = randint(10, 15)
        b_val = randint(10, 15)
        c_val = randint(10, 15)
        
        # Check triangle inequality
        if a_val + b_val > c_val and a_val + c_val > b_val and b_val + c_val > a_val:
            s = Rational(a_val + b_val + c_val, 2)
            area_squared = s * (s - a_val) * (s - b_val) * (s - c_val)
            
            if area_squared > 0:
                area_val = sqrt(area_squared)
                
                question = f"Find the area of a triangle with sides of length ${a_val}$, ${b_val}$, and ${c_val}$ using Heron's formula."
                
                ans = area_val
                
                solution = steps(
                    f"Heron's formula: Area $= \\sqrt{{s(s-a)(s-b)(s-c)}}$ where $s$ is the semi-perimeter",
                    f"Calculate semi-perimeter: $s = \\frac{{{a_val} + {b_val} + {c_val}}}{{2}} = {latex(s)}$",
                    f"$s - a = {latex(s)} - {a_val} = {latex(s - a_val)}$",
                    f"$s - b = {latex(s)} - {b_val} = {latex(s - b_val)}$",
                    f"$s - c = {latex(s)} - {c_val} = {latex(s - c_val)}$",
                    f"Area $= \\sqrt{{{latex(s)} \\cdot {latex(s - a_val)} \\cdot {latex(s - b_val)} \\cdot {latex(s - c_val)}}}$",
                    f"Area $= \\sqrt{{{latex(area_squared)}}}$",
                    f"Area $= {latex(area_val)}$"
                )
                
                return problem(
                    question=question,
                    answer=ans,
                    difficulty=(1700, 1800),
                    topic="geometry/triangle_properties",
                    solution=solution
                )
    
    # Fallback to angle bisector problem
    ratio1 = 3
    ratio2 = 4
    multiplier = 2
    side1 = ratio1 * multiplier
    side2 = ratio2 * multiplier
    total_base = 14
    segment1 = Rational(total_base * ratio1, ratio1 + ratio2)
    
    question = f"In triangle $ABC$, $AC = {side1}$ and $BC = {side2}$. The angle bisector from $C$ meets side $AB$ at point $D$, dividing it into segments $AD$ and $DB$ where $AD + DB = {total_base}$. Find $AD$."
    
    ans = segment1
    
    solution = steps(
        f"By the Angle Bisector Theorem: $\\frac{{AD}}{{DB}} = \\frac{{AC}}{{BC}} = \\frac{{{side1}}}{{{side2}}}$",
        f"Let $AD = x$ and $DB = y$, so $x + y = {total_base}$",
        f"From the ratio: $\\frac{{x}}{{y}} = \\frac{{{side1}}}{{{side2}}}$, so $y = \\frac{{{side2}x}}{{{side1}}}$",
        f"Substituting: $x + \\frac{{{side2}x}}{{{side1}}} = {total_base}$",
        f"$x = {latex(ans)}$"
    )
    
    return problem(
        question=question,
        answer=ans,
        difficulty=(1650, 1750),
        topic="geometry/triangle_properties",
        solution=solution
    )

emit(generate())

File: src/geometry_triangle_properties_hard.jl
# geometry - triangle_properties (hard)
# Generated: 2026-03-08T20:03:02.080979

include(joinpath(@__DIR__, "..", "julia", "src", "ProblemUtils.jl"))
using .ProblemUtils

@script begin
    set_topic!("geometry/triangle_properties")
    
    problem_type = choice(1:5)
    
    if problem_type == 1
        # Triangle with angle bisector theorem (ELO 1800-2000)
        AB = randint(8, 20)
        AC = randint(8, 20)
        BD = randint(3, 10)
        DC = (AC * BD) // AB
        BC = BD + DC
        
        d = DiagramObj(600, 400)
        point!(d, 100, 300, "A")
        point!(d, 500, 300, "B")
        point!(d, 300, 50, "C")
        point!(d, 300 + (200 * BD) / BC - 200, 300, "D")
        
        line!(d, 100, 300, 500, 300)
        line!(d, 100, 300, 300, 50)
        line!(d, 500, 300, 300, 50)
        line!(d, 300, 50, 300 + (200 * BD) / BC - 200, 300, color="blue", stroke_width=2)
        
        segment_label!(d, 100, 300, 500, 300, "$(BC)", :below)
        segment_label!(d, 100, 300, 300, 50, "$(AB)", :left)
        segment_label!(d, 500, 300, 300, 50, "$(AC)", :right)
        text!(d, 300 + (200 * BD) / BC - 200, 320, "D", font_size=16)
        segment_label!(d, 100, 300, 300 + (200 * BD) / BC - 200, 300, "$(BD)", :below)
        
        problem(
            question="In triangle ABC, AD is the angle bisector of ∠BAC. If AB = $(AB), AC = $(AC), and BD = $(BD), find DC.",
            answer=DC,
            difficulty=(1800, 2000),
            solution=steps(
                "By the Angle Bisector Theorem, \\frac{BD}{DC} = \\frac{AB}{AC}",
                sol("Substituting", "\\frac{$(BD)}{DC} = \\frac{$(AB)}{$(AC)}"),
                sol("Cross multiply", "$(AB) \\cdot DC = $(AC) \\cdot $(BD) = $(AC * BD)"),
                sol("DC", DC)
            ),
            image=render(d),
            time=120
        )
        
    elseif problem_type == 2
        # Stewart's Theorem application (ELO 2000-2200)
        a = randint(10, 18)
        b = randint(10, 18)
        m = randint(4, 9)
        n = randint(4, 9)
        c = m + n
        
        d_squared = (b^2 * m + a^2 * n - m * n * c) // c
        d_val = sqrt(Float64(d_squared))
        
        d = DiagramObj(600, 400)
        point!(d, 100, 300, "B")
        point!(d, 500, 300, "C")
        point!(d, 300, 50, "A")
        point!(d, 100 + 400 * m / c, 300, "D")
        
        line!(d, 100, 300, 500, 300)
        line!(d, 100, 300, 300, 50)
        line!(d, 500, 300, 300, 50)
        line!(d, 300, 50, 100 + 400 * m / c, 300, color="blue", stroke_width=2)
        
        segment_label!(d, 100, 300, 300, 50, "$(a)", :left)
        segment_label!(d, 500, 300, 300, 50, "$(b)", :right)
        segment_label!(d, 100, 300, 100 + 400 * m / c, 300, "$(m)", :below)
        segment_label!(d, 100 + 400 * m / c, 300, 500, 300, "$(n)", :below)
        text!(d, 100 + 400 * m / c, 320, "D", font_size=16)
        
        problem(
            question="In triangle ABC, point D lies on BC such that BD = $(m) and DC = $(n). If AB = $(a) and AC = $(b), find AD² using Stewart's Theorem.",
            answer=d_squared,
            difficulty=(2000, 2200),
            solution=steps(
                "Stewart's Theorem states: b²m + a²n = c(d² + mn)",
                "Where BC = c = $(m) + $(n) = $(c) and AD = d",
                sol("Substitute", "$(b)^2 \\cdot $(m) + $(a)^2 \\cdot $(n) = $(c)(d^2 + $(m) \\cdot $(n))"),
                sol("Simplify", "$(b^2 * m) + $(a^2 * n) = $(c)d^2 + $(c * m * n)"),
                sol("Solve for d²", "d^2 = \\frac{$(b^2 * m + a^2 * n) - $(c * m * n)}{$(c)} = $(d_squared)")
            ),
            image=render(d),
            time=180
        )
        
    elseif problem_type == 3
        # Median length formula (ELO 1900-2100)
        a = randint(8, 16)
        b = randint(8, 16)
        c = randint(8, 16)
        
        m_a_squared = (2*b^2 + 2*c^2 - a^2) // 4
        
        d = DiagramObj(600, 400)
        point!(d, 100, 300, "B")
        point!(d, 500, 300, "C")
        point!(d, 300, 50, "A")
        point!(d, 300, 300, "M")
        
        line!(d, 100, 300, 500, 300)
        line!(d, 100, 300, 300, 50)
        line!(d, 500, 300, 300, 50)
        line!(d, 300, 50, 300, 300, color="red", stroke_width=2)
        
        segment_label!(d, 100, 300, 500, 300, "a = $(a)", :below)
        segment_label!(d, 100, 300, 300, 50, "c = $(c)", :left)
        segment_label!(d, 500, 300, 300, 50, "b = $(b)", :right)
        text!(d, 300, 320, "M", font_size=16)
        
        problem(
            question="In triangle ABC, M is the midpoint of BC. If BC = $(a), AC = $(b), and AB = $(c), find the square of the length of median AM using the formula m_a² = (2b² + 2c² - a²)/4.",
            answer=m_a_squared,
            difficulty=(1900, 2100),
            solution=steps(
                "The median length formula is m_a^2 = \\frac{2b^2 + 2c^2 - a^2}{4}",
                sol("Substitute", "m_a^2 = \\frac{2($(b))^2 + 2($(c))^2 - ($(a))^2}{4}"),
                sol("Calculate", "m_a^2 = \\frac{$(2*b^2) + $(2*c^2) - $(a^2)}{4}"),
                sol("Answer", "m_a^2 = \\frac{$(2*b^2 + 2*c^2 - a^2)}{4} = $(m_a_squared)")
            ),
            image=render(d),
            time=150
        )
        
    elseif problem_type == 4
        # Triangle with altitude and area relationship (ELO 1850-2050)
        base = randint(12, 24)
        height = randint(8, 18)
        area = (base * height) // 2
        
        side_a = randint(10, 20)
        altitude_to_a = (2 * area) // side_a
        
        d = DiagramObj(600, 400)
        point!(d, 100, 300, "B")
        point!(d, 500, 300, "C")
        point!(d, 300, 50, "A")
        point!(d, 300, 300, "H")
        
        line!(d, 100, 300, 500, 300)
        line!(d, 100, 300, 300, 50)
        line!(d, 500, 300, 300, 50)
        line!(d, 300, 50, 300, 300, color="green", stroke_width=2, stroke_dasharray="5,5")
        right_angle!(d, 300, 300, 300, 50, 350, 300, size=15)
        
        segment_label!(d, 100, 300, 500, 300, "$(base)", :below)
        segment_label!(d, 300, 50, 300, 300, "$(height)", :right)
        text!(d, 300, 320, "H", font_size=16)
        
        problem(
            question="Triangle ABC has base BC = $(base) and altitude AH = $(height) from A to BC. If another side of the triangle has length $(side_a), what is the length of the altitude to that side?",
            answer=altitude_to_a,
            difficulty=(1850, 2050),
            solution=steps(
                sol("Area of triangle", "A = \\frac{1}{2} \\cdot $(base) \\cdot $(height) = $(area)"),
                "Using the same area with the other side as base:",
                sol("Formula", "$(area) = \\frac{1}{2} \\cdot $(side_a) \\cdot h"),
                sol("Solve", "h = \\frac{2 \\cdot $(area)}{$(side_a)} = \\frac{$(2*area)}{$(side_a)} = $(altitude_to_a)")
            ),
            image=render(d),
            time=120
        )
        
    else
        # Angle bisector with perpendiculars (ELO 2100-2300)
        angle_deg = choice([30, 40, 50, 60])
        half_angle = angle_deg / 2
        dist = randint(8, 16)
        
        perpendicular_dist = dist * sin(deg2rad(half_angle))
        rounded_ans = round(perpendicular_dist, digits=2)
        
        d = DiagramObj(600, 500)
        point!(d, 100, 400, "A")
        point!(d, 500, 400, "B")
        point!(d, 100, 100, "C")
        point!(d, 300, 250, "P")
        
        line!(d, 100, 400, 500, 400)
        line!(d, 100, 400, 100, 100)
        line!(d, 100, 400, 300, 250, color="blue", stroke_width=2)
        line!(d, 300, 250, 300, 400, color="red", stroke_width=1.5, stroke_dasharray="5,5")
        line!(d, 300, 250, 100, 250, color="red", stroke_width=1.5, stroke_dasharray="5,5")
        right_angle!(d, 300, 400, 300, 250, 350, 400, size=12)
        right_angle!(d, 100, 250, 300, 250, 100, 300, size=12)
        
        angle_arc!(d, 100, 400, 500, 400, 100, 100, radius=50)
        text!(d, 170, 380, "$(angle_deg)°", font_size=14)
        segment_label!(d, 100, 400, 300, 250, "$(dist)", :above)
        text!(d, 300, 420, "Q", font_size=16)
        text!(d, 80, 250, "R", font_size=16)
        
        problem(
            question="Point P lies on the angle bisector of ∠BAC where ∠BAC = $(angle_deg)°. If AP = $(dist), find the perpendicular distance from P to side AB. Round to 2 decimal places.",
            answer=rounded_ans,
            difficulty=(2100, 2300),
            solution=steps(
                "Since AP is on the angle bisector, the angle from AB to AP is $(half_angle)°",
                "The perpendicular distance d from P to AB satisfies \\sin($(half_angle)°) = \\frac{d}{$(dist)}",
                sol("Calculate", "d = $(dist) \\cdot \\sin($(half_angle)°)"),
                sol("Answer", "d \\approx $(rounded_ans)")
            ),
            image=render(d),
            calculator="scientific",
            time=180
        )
    end
end